src/dag/lazy-store.ts

```diff
@@ -0,0 +1,350 @@
+import {Hash, isTempHash} from '../hash';
+import type * as kv from '../kv/mod';
+import {RWLock} from '../rw-lock';
+import {Chunk, ChunkHasher, createChunk} from './chunk';
+import type {Store, Read, Write} from './store';
+import {getSizeOfValue as defaultGetSizeOfValue} from '../get-size-of-value';
+import type {ReadonlyJSONValue} from '../mod';
+import {GarbargeCollector, HeadChange} from './gc';
+
+export class LazyStore implements Store {
+  private readonly _rwLock = new RWLock();
```

Comment:
I think it's nice to add a comment saying what state/fields a lock is protecting.

Reply:
Side note: it would be neat to have a kind of lock that self-documents that, like:

```
RwLock<Stuff>
```
And then when you do `withLock` you receive the stuff as a param.

Reply:
but where does stuff come from? the RWLock callaback have no params.

Reply:
I'm the big picture guy @arv 😂.

I dunno, I was thinking maybe  new lock class that takes the thingy to lock as a parameter. Like:

```
const lock = new FancyLock({foo: "bar"});
lock.withLock(async stuff => {
  // do stuff with stuff
});
```

Reply:
We can do that but that is a higher abstraction.

Reply:
added a comment

```
  /**
   * This lock is used to ensure correct isolation of Reads and Writes.  Multiple Reads
   * are allowed in parallel but only a single Write.  Reads and Writes see a isolated
   * view of the store (corresponding to the Serializable level of transaction isolation
   * defined in the SQL standard).
   *
   * To ensure these semantics the read lock must be acquired when a Read is created
   * and held til it is closed, and a Write lock must be acquired when a Write is
   * created and held til it is commited.
   *
   * Code must have a read or write lock to
   * - read `heads`
   * - read `tempChunks`
   * - read `_sourceStore`
   * - read and write `_sourceChunksCache`
   * - read and write `_refCounts`
   * and must have a write lock to
   * - write `heads`
   * - write `tempChunks`
   */
```